### ISSUE TITLE
Nesting keras.Models makes tf.gradients() not update while training

### ISSUE THREAD

**Issue by fotlogo** [tensorflow/tensorflow#58624]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.10

### Custom Code

No

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.8

### Current Behaviour?

```shell
Nesting a keras.model  inside another one makes analytic gradients computed with tf.gradients() not update with training. On the toy example presented bellow, the inside network consists of a single multiplication with a trainable variable (i.e. Dense(1, use_bias=False)), computes derivative, and concatenas all 3 so  for input [x] output should be [x,a*x,a] (a is the traibale variable). Training this independently works as expected. 
But if you nest that inside another keras.model(), even a trivial one that multiplies *2, devides /2 (i.e. no change) and then calls the inside model, the gradient computation is unchanged after training. so after you train, the output is [x,a*x,a0] where a0 is the initial weight (1.5 vs 2 in my toy example). 
Computing gradients with tf.GradientTape as in here also has the exact same behaviour  https://www.tensorflow.org/guide/advanced_autodiff

My 'real world' project involved solving a differential equation using https://github.com/titu1994/tf_SIREN  and switching bewteen training inside and outside model. As gradients dont work properly this way, my workaround was to inline the code of inside model to outside, keep 2 model instances and manual copy weights from one to the other which feels very hacky and I am sure is not indeded.

Tested on tf 2.6,2.8,2.10 , both Ubuntu/Windows and the behaviour is identical everywhere.
```

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf
tf.random.set_seed(0)

from tensorflow.keras.optimizers import Adam,SGD
from tensorflow.keras import Input, Model,Sequential
from tensorflow.keras.initializers import Constant
from tensorflow.keras.layers import Layer, Dense
from tensorflow.keras.layers import Concatenate, concatenate


tf.compat.v1.disable_eager_execution()

#trivial network doing multiplication by a single trainable variable
def inside_network():
    x = Input(1)    
        
    layer=Dense(1,activation='linear',use_bias=False, kernel_initializer=Constant(value=1.5)) 
    y=layer(x)    

    dy=tf.gradients(y,[x])[0]  
    #for weight a, output should be [x, ax, a]
    out= concatenate([x, y, dy],axis=-1)
    
    network = Model(inputs = x, outputs = out, name='inside_network')
    
    return network

def outside_network():
    x = Input((1))      
    y=2*x
    y=y*0.5
    model=inside_network()
    y=model(y)
    
    outside_network = Model(inputs = x, outputs = y, name='outside_network')
    return outside_network
    
def get_first_weights(mod):
    for layer in mod.layers:
        w=layer.get_weights()
        if len(w)>0:
            return w

Ndata=1024
np.random.seed(0)
xdata=np.random.rand(Ndata,1)
xdata[0:4,0]=np.arange(0,4)
a=2.0

data_train_target=np.concatenate([xdata, a*xdata, a*np.ones((Ndata,1))],axis=-1)
print('input :4',xdata[:4,:])
print('target :4',data_train_target[:4,:])

model1=inside_network()

model1.compile(optimizer=Adam(1e-2),  loss='mae') #
print('dense layer weights before training',get_first_weights(model1))

model1.fit(xdata,data_train_target,verbose=1,epochs=2,shuffle=True) #,batch_size=bs
   
print('dense layer weights after training',get_first_weights(model1))
           
dataout=model1.predict(xdata)
print('dataout',dataout[0:4])
print('target',data_train_target[0:4,:])

model2=outside_network()

model2.compile(optimizer=Adam(1e-2),  loss='mae') #

print('dense layer weights before training',get_first_weights(model2))



model2.fit(xdata,data_train_target,verbose=1,epochs=2,shuffle=True)

print('dense layer weights after training',get_first_weights(model2))

dataout2=model2.predict(xdata)
print('dataout2',dataout2[0:4])
print(data_train_target[:4,:])
```

### Relevant log output

```shell
input :4 [[0.]
 [1.]
 [2.]
 [3.]]
target :4 [[0. 0. 2.]
 [1. 2. 2.]
 [2. 4. 2.]
 [3. 6. 2.]]
2022-11-20 18:27:50.249131: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:50.295560: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:50.295870: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:50.298148: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:50.298388: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:50.298603: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:51.014096: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:51.014341: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:51.014531: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-20 18:27:51.014710: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 8946 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:09:00.0, compute capability: 7.5
dense layer weights before training [array([[1.5]], dtype=float32)]
Train on 1024 samples
Epoch 1/2
1024/1024 [==============================] - 1s 927us/sample - loss: 0.1731
Epoch 2/2
1024/1024 [==============================] - 0s 43us/sample - loss: 0.0320
dense layer weights after training [array([[2.011962]], dtype=float32)]
/home/flogothetis/.local/lib/python3.8/site-packages/keras/engine/training_v1.py:2079: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,
dataout [[0.       0.       2.011962]
 [1.       2.011962 2.011962]
 [2.       4.023924 2.011962]
 [3.       6.035886 2.011962]]
target [[0. 0. 2.]
 [1. 2. 2.]
 [2. 4. 2.]
 [3. 6. 2.]]
dense layer weights before training [array([[1.5]], dtype=float32)]
Train on 1024 samples
Epoch 1/2
1024/1024 [==============================] - 0s 142us/sample - loss: 0.2240
Epoch 2/2
1024/1024 [==============================] - 0s 58us/sample - loss: 0.1774
dense layer weights after training [array([[2.0133402]], dtype=float32)]
dataout2 [[0.        0.        1.5      ]
 [1.        2.0133402 1.5      ]
 [2.        4.0266805 1.5      ]
 [3.        6.040021  1.5      ]]
[[0. 0. 2.]
 [1. 2. 2.]
 [2. 4. 2.]
 [3. 6. 2.]]
```
</details>

**Comment by svobora** [comment]
Your inside network is not connected to the outside bro, because you have 2 inputs.

**Comment by svobora** [comment]
[tf.gradients](https://www.tensorflow.org/api_docs/python/tf/gradients) is only valid in a graph context. In particular, it is valid in the context of a [tf.function](https://www.tensorflow.org/api_docs/python/tf/function) wrapper, where code is executing as a graph.

**Comment by tilakrayal** [comment]
@fotlogo,
Please take a look at above comment and also xs and ys are each a Tensor or a list of tensors. grad_ys is a list of Tensor, holding the gradients received by the ys. The list must be the same length as ys.

gradients() adds ops to the graph to output the derivatives of ys with respect to xs. It returns a list of Tensor of length len(xs) where each tensor is the sum(dy/dx) for y in ys and for x in xs. Thank you!

**Comment by fotlogo** [comment]
@svobora thank for the very quick reply! I was under the impression that including layer in a keras.model()  and then calling model.compile() and model.fit() guarantees that the code will run in graph mode. and in any case I have included the 'tf.compat.v1.disable_eager_execution() '  to make sure I only run in graph mode right ?
@tilakrayal there should not be any sizing issues as the code runs fine (otherwise it will have generated a run time error) and indeed the fist part of the code training the inside_network() works exactly as expected  with the gradient being computed correctly (see the first part of my output)
the problem arises when you nest that code inside another model, and then the gradients are still computed but not being updated after the  training (final collumn of output being 1.5 were it should be 2)

**Comment by svobora** [comment]
I think you are making a mistake in treatment of a Network (a sequence of tf layers) and a Model (an object to handle fitting). Just use layer sequences to form networks, then build a single Model on top of the however interconnected networks.

```
import numpy as np
import tensorflow as tf

tf.random.set_seed(0)

from tensorflow.keras.optimizers import Adam, SGD
from tensorflow.keras import Input, Model, Sequential
from tensorflow.keras.initializers import Constant
from tensorflow.keras.layers import Layer, Dense
from tensorflow.keras.layers import Concatenate, concatenate

tf.compat.v1.disable_eager_execution()


# trivial network doing multiplication by a single trainable variable
def inside_network(x):

    layer = Dense(1, activation='linear', use_bias=False, kernel_initializer=Constant(value=1.5))
    y = layer(x)

    dy = tf.gradients(y, [x])[0]
    # for weight a, output should be [x, ax, a]
    out = concatenate([x, y, dy], axis=-1)

    return out


def outside_network():
    x = Input((1))
    y = 2 * x
    y = y * 0.5
    y = inside_network(y)

    outside_network = Model(inputs=x, outputs=y, name='outside_network')
    return outside_network


def get_first_weights(mod):
    for layer in mod.layers:
        w = layer.get_weights()
        if len(w) > 0:
            return w


Ndata = 1024
np.random.seed(0)
xdata = np.random.rand(Ndata, 1)
xdata[0:4, 0] = np.arange(0, 4)
a = 2.0

data_train_target = np.concatenate([xdata, a * xdata, a * np.ones((Ndata, 1))], axis=-1)
print('input :4', xdata[:4, :])
print('target :4', data_train_target[:4, :])

x = Input((1))
model1 = Model(x, inside_network(x), name='inside_network')

model1.compile(optimizer=Adam(1e-2), loss='mae')  #
print('dense layer weights before training', get_first_weights(model1))

model1.fit(xdata, data_train_target, verbose=1, epochs=2, shuffle=True)  # ,batch_size=bs

print('dense layer weights after training', get_first_weights(model1))

dataout = model1.predict(xdata)
print('dataout', dataout[0:4])
print('target', data_train_target[0:4, :])

model2 = outside_network()

model2.compile(optimizer=Adam(1e-2), loss='mae')  #

print('dense layer weights before training', get_first_weights(model2))

model2.fit(xdata, data_train_target, verbose=1, epochs=2, shuffle=True)

print('dense layer weights after training', get_first_weights(model2))

dataout2 = model2.predict(xdata)
print('dataout2', dataout2[0:4])
print(data_train_target[:4, :])
```

**Comment by tilakrayal** [comment]
@fotlogo,
I was able to reproduce the issue on tensorflow v2.10, whereas on v2.11 I was facing a different issue. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/6b56b0071731ecd93ba446786f77037b/untitled786.ipynb). Thank you!

**Comment by fotlogo** [comment]
@tilakrayal Sorry for the late reply ! According to the changelog in https://github.com/tensorflow/tensorflow/releases 'old' optimisers on 2.11 are moved in tensorflow.keras.optimizers.legacy . so by changing the import as from tensorflow.keras.optimizers.legacy import Adam,SGD the exact same behaviour as in 2.10 is observed.
updated the gist [here ](https://colab.research.google.com/gist/fotlogo/34cf3a1fef8f70c454b2243ac7e14760/untitled786.ipynb)

**Comment by tilakrayal** [comment]
@SuryanarayanaY,
I was able to reproduce the issue on tensorflow v2.10, v2.11 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/aee0f47445cac1b9967d4f39a2bdfe1b/untitled786.ipynb).

**Comment by SuryanarayanaY** [comment]
Hi @fotlogo ,

Let me explain my understanding here. For model1 (inside_network) you have initialized weight as 1.5. Then you passed `inputs = 'x'` and `outputs = [x,ax,a]` where `a` is the derivative/weight.The value of 'a' before training is 1.5 here. After training model1 the 'a' value becomes 2.011962.

In second model i.e `model2` (outside network) the `inputs ='x'` and the `outputs=inside_network()(y)`, here inside_network() is not a trained model yet and the value of outputs is [x,ax,a]. Note here the last value 'a' is initialized as 1.5 but when passed it to second model 'a' is just a constant value only and the value 'ax' is the variable and here only the coefficient 'a' value getting trained and updated to `2.0133402` from `1.5`.For model2 the last value in concat layer is just a fixed constant i.e. `1.5` which will not update.But the value `a` in 'ax' updated to `2.0133402` and you can see the predict values multiples of this except last column which is a constant '1.5'.

I hope this explanation will give some clarity on why the value `1.5` not getting updates in model2 .Please feel free to come back if still have questions.

Thanks!

**Comment by SuryanarayanaY** [comment]
Closing as stale. Please reopen if you'd like to work on this further. Thanks.
